### PR TITLE
Add support for Base64 Encode/Decode Scalar Functions

### PR DESCRIPTION
**Pull request by SabrinaZhaozyf**
Label = `feature`

This PR implements `toBase64(byte[])` and `fromBase64(str)` encoding and decoding functionality.
- Encoding scheme: encodes input using the Base64 encoding scheme as specified in RFC 4648 and RFC 2045
- Decoding will throw an exception if invalid base64 string is passed in.

Use Case Examples
- Encode binary data: `toBase64(binaryCol)` 
- Encode strings in UTF8: `toBase64(toUtf8('hello'))` , `toBase64(toUtf8(stringCol))` 
- Decode base64-encoded strings to binary data: `fromBase64('aGVsbG8h')`, `fromBase64(stringCol)`
- Decode base64-encoded strings to UTF8 charset: `fromUtf8(fromBase64('aGVsbG8h'))`

### CODE REVIEW

**Comment by amrishlal** [comment]
In the description, can we add a short overview of the issue being addressed along with a summary of changes.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 1]
```diff
@@ -74,13 +74,9 @@ public void testCanonicalFunctionName() {
...
     //@formatter:off
     PinotQuery pinotQuery = CalciteSqlParser.compileToPinotQuery(
-        "SELECT OrderID, Quantity,\n"
-            + "CASE\n"
-            + "    WHEN Quantity > 30 THEN 'The quantity is greater than 30'\n"
-            + "    WHEN Quantity = 30 THEN 'The quantity is 30'\n"
-            + "    ELSE 'The quantity is under 30'\n"
-            + "END AS QuantityText\n"
-            + "FROM OrderDetails");
+        "SELECT OrderID, Quantity,\n" + "CASE\n" + "    WHEN Quantity > 30 THEN 'The quantity is greater than 30'\n"
```
Comment (amrishlal) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:77]: This change appears to be unrelated to the issue, can we revert this any other similar changes?

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 2]
```diff
@@ -560,4 +561,26 @@ public static String decodeUrl(String input)
       throws UnsupportedEncodingException {
     return URLDecoder.decode(input, StandardCharsets.UTF_8.toString());
   }
+
+  /**
+   *
+   * @param input utf-8 encoded String
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:567]: Java String object is not UTF-8. This should probably just say `@param input input string`.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 3]
```diff
@@ -560,4 +561,26 @@ public static String decodeUrl(String input)
       throws UnsupportedEncodingException {
     return URLDecoder.decode(input, StandardCharsets.UTF_8.toString());
   }
+
+  /**
+   *
```
Comment (amrishlal) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:566]: minor: I would get rid of the extra blank line in the comment to save vertical space on the screen.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 4]
```diff
@@ -560,4 +561,24 @@ public static String decodeUrl(String input)
       throws UnsupportedEncodingException {
     return URLDecoder.decode(input, StandardCharsets.UTF_8.toString());
   }
+
+  /**
+   * @param input utf-8 encoded String
+   * @return Base64 encoded String
+   */
+  @ScalarFunction
+  public static String toBase64(String input) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:570]: We should decouple the functionality into 2 functions:
1. encode/decode base64
2. encode/decode utf8

Currently we are mixing the 2, which is quite confusing, and not as flexible

File: pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java [hunk 5]
```diff
@@ -211,6 +219,59 @@ public void testBrokerRequestHandlerWithAsFunction()
     Assert.assertEquals(brokerResponse.getResultTable().getRows().get(0)[1].toString(),
         "key1=value 1&key2=value@!$2&key3=value%3");
     Assert.assertEquals(brokerResponse.getTotalDocs(), 0);
+
+    request = JsonUtils.stringToJsonNode(
+        "{\"sql\":\"SELECT toBase64('hello!') AS encoded, " + "fromBase64('aGVsbG8h') AS decoded\"}");
+    requestStats = Tracing.getTracer().createRequestScope();
+    brokerResponse = requestHandler.handleRequest(request, null, requestStats);
+    System.out.println(brokerResponse.getResultTable());
```
Comment (siddharthteotia) [inline pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java:227]: (nit) this code can be made slightly more readable imho by doing get on the resultTable , dataSchema and getRows exactly once.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 6]
```diff
@@ -560,4 +561,24 @@ public static String decodeUrl(String input)
       throws UnsupportedEncodingException {
     return URLDecoder.decode(input, StandardCharsets.UTF_8.toString());
   }
+
+  /**
+   * @param input utf-8 encoded String
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:566]: Suggest changing this to `input string to be encoded`

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 7]
```diff
@@ -1890,6 +1890,41 @@ public void testCompilationInvokedFunction() {
     decoded = and.getOperands().get(1).getFunctionCall().getOperands().get(1).getLiteral().getStringValue();
     Assert.assertEquals(encoded, "key1%3Dvalue+1%26key2%3Dvalue%40%21%242%26key3%3Dvalue%253");
     Assert.assertEquals(decoded, "key1=value 1&key2=value@!$2&key3=value%3");
+
+    query = "select toBase64('hello!'), fromBase64('aGVsbG8h') from mytable";
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:1894]: Can you add tests for nested transform as well ?

File: pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java [hunk 8]
```diff
@@ -104,6 +104,12 @@ public void testLiteralOnlyTransformBrokerRequestFromSQL() {
     Assert.assertFalse(BaseBrokerRequestHandler.isLiteralOnlyQuery(CalciteSqlParser
         .compileToPinotQuery("SELECT count(*) from foo "
             + "where bar = decodeUrl('key1%3Dvalue+1%26key2%3Dvalue%40%21%242%26key3%3Dvalue%253')")));
+    Assert.assertTrue(BaseBrokerRequestHandler.isLiteralOnlyQuery(
+        CalciteSqlParser.compileToPinotQuery("SELECT toBase64('hello!')," + " fromBase64('aGVsbG8h') FROM myTable")));
+    Assert.assertFalse(BaseBrokerRequestHandler.isLiteralOnlyQuery(
```
Comment (siddharthteotia) [inline pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java:109]: Can you add a test for nested transform which does not result in scalar / literal only resolution ?

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 7 above)
Comment (siddharthteotia) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:1894]: - toBase64 is inside another transform function that produces string
- toBase64 is inside fromBase64
- etc ...

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 4 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:570]: Hi Jackie, thanks for the input!
I decided to set the argument type to String as that follows [MySql’s syntax](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_to-base64) and will give better usability.
`TO_BASE64(str)` will use default JVM encoding.

However, I do understand where you are coming from and there are 2 things that I want to propose for better flexibility.
`TO_BASE64(str, encodingscheme)` will use user-specified encoding scheme.
`BINARY_TO_BASE64(byte[])` directly encodes binary to Base64.

Please let me know what you think!

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:580]: I feel the [presto syntax](https://prestodb.io/docs/current/functions/binary.html) is easier to use though because base64 is essentially a way to encode binary to string. Noted that in the MySQL case, it uses the character set from the connection, which is not available in pinot

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 9]
```diff
@@ -560,4 +562,42 @@ public static String decodeUrl(String input)
...
     return URLDecoder.decode(input, StandardCharsets.UTF_8.toString());
   }
+
+  /**
+   * @param input binary data
+   * @return Base64 encoded String
+   */
+  @ScalarFunction
+  public static String binaryToBase64(byte[] input) {
+    return Base64.getEncoder().encodeToString(input);
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:572]: What is the expectation on charset of input bytes / binary data ? Looks like we assume it is UTF8. What if the input bytes provided here was encoded using UTF16

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 10]
```diff
@@ -560,4 +562,42 @@ public static String decodeUrl(String input)
...
+    byte[] inputBytes = input.getBytes(Charset.forName(encodeScheme));
+    return Base64.getEncoder().encodeToString(inputBytes);
+  }
+
+  /**
+   * @param input Base64 encoded String
+   * @return decoded string
+   */
+  @ScalarFunction
+  public static String fromBase64(String input) {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:599]: If we are providing a toBase64 with a user specified charset encoding and a default utf8 charset flavor then shouldn't we also provide fromBase64 in both flavors ?

What if someone calls `toBase64("blahblah", "UTF16")` to get "something" and then `fromBase64("something")` - this will convert "something" to bytes and then to String using UTF8 which can be wrong because "something" was created from "blahblah" using UTF8.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 4 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:580]: Thanks for the followup! After considering both you and @siddharthteotia 's comments, I propose the following two approaches. Please let me know what you think.

**Method 1 (byte[] to byte[] for both TO_BASE64 and FROM_BASE64)**

- TO_BASE64: byte[] -> base64 encoding -> byte[] =====> user can convert output byte[] to any charset they choose (need to add support for fromASCII and fromUTF8)

- user can convert input string to byte[] using any charset they choose (currently pinot supports toASCII and toUTF8) =====> FROM_BASE64: byte[] -> base64 decoding -> byte[]

**Method 2 (byte[] to String and String to byte[] using user-specified charset)**

- TO_BASE64: byte[] -> base64 encoding -> byte[] -----(user-specified charset)-----> String

- FROM_BASE64: String ----(user-specified charset)--->  byte[] -> base64 decoding -> byte[]

cc: @siddharthteotia @jasperjiaguo

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 2 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java]: Resolved - updated implementation passes byte[] as argument instead of String.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 3 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java]: Done

File: pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java (hunk 5 above)
Comment (SabrinaZhaozyf) [reply pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java]: Done!

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 6 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java]: Resolved - updated implementation passes byte[] as argument instead of String.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 7 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java]: Done - added nested function calls with other transform functions(eg. upper, reverse) both in SELECT and in filter.

File: pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java (hunk 8 above)
Comment (SabrinaZhaozyf) [reply pinot-broker/src/test/java/org/apache/pinot/broker/requesthandler/LiteralOnlyBrokerRequestTest.java:117]: Done - added nested transform function that takes in column as an argument.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 9 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:580]: Resolved - we will not have a fixed charset of input bytes now with the new implementation. For example, if user wants to encode the String "hello!", he/she can choose to do `toBase64(toUtf8("hello"))` or `toBase64(toUtf16("hello"))` (Pinot currently only supports toUtf8 and toASCII functions). During decoding, he/she can do `fromUtf8(fromBase64("fvduivheui"))` to `fromUtf16(fromBase64("fvduivheui"))` to get "hello!" depending on if they used UTF8 or UTF16 when they encoded.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 10 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java]: Resolved - `fromBase64` will decode into byte[] (the layer from byte[] to String is removed now) and the user can decide to encode the byte[] to UTF8 strings or other charsets by calling other functions such as `fromUtf8`.

**Review by siddharthteotia — APPROVED** [review]
Thanks for the contribution

**Comment by siddharthteotia** [comment]
@SabrinaZhaozyf - please update the PR description to reflect the latest discussion / approach

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java [hunk 11]
```diff
@@ -393,6 +394,15 @@ public static byte[] toUtf8(String input) {
     return input.getBytes(StandardCharsets.UTF_8);
   }
 
+  /**
+   * @param input
```
Comment (jasperjiaguo) [inline pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:398]: complete the java doc here?

**Review by jasperjiaguo — CHANGES_REQUESTED** [review]
Minor questions regarding invalid input / unexpected usage.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java [hunk 12]
```diff
@@ -600,6 +599,173 @@ public void testUrlFunc()
...
+    for (int i = 0; i < 100; i++) {
+      String original = rows.get(0).asText();
+      String encoded = rows.get(1).asText();
+      String decoded = rows.get(2).asText();
+      assertEquals(original, decoded);
+      assertEquals(encoded, toBase64(toUtf8(original)));
+      assertEquals(decoded, fromUtf8(fromBase64(toBase64(toUtf8(original)))));
+    }
+
+    // invalid argument
```
[the remark this answers is not in the document]
Comment (SabrinaZhaozyf) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:674]: Done! `toBase64('hello!')`will error out during compilation.

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 11 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java]: Done.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (SabrinaZhaozyf) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/OfflineClusterIntegrationTest.java:674]: Will add in documentation that BYTES col will be represented as HEX string (also link: https://docs.pinot.apache.org/users/user-guide-query/querying-pinot#bytes-column)

File: pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java (hunk 9 above)
Comment (SabrinaZhaozyf) [reply pinot-common/src/main/java/org/apache/pinot/common/function/scalar/StringFunctions.java:580]: Resolved - added tests for calling toBase64 on a string. Will add in documentation that BYTES col will be represented in HEX string(also link:https://docs.pinot.apache.org/users/user-guide-query/querying-pinot#bytes-column).

**Review by jasperjiaguo — APPROVED** [review]

**Comment by siddharthteotia** [comment]
@SabrinaZhaozyf - let's user docs to gitbook please

**Comment by npawar** [comment]
> @SabrinaZhaozyf - let's user docs to gitbook please

Hello, was this added into the docs? I cannot find it in https://docs.pinot.apache.org/configuration-reference/functions

**Comment by SabrinaZhaozyf** [comment]
@npawar Changes to user doc has been merged here (https://github.com/pinot-contrib/pinot-docs/pull/109). Not sure why it hasn't been updated on documentation yet.

**Comment by npawar** [comment]
> @npawar Changes to user doc has been merged here ([pinot-contrib/pinot-docs#109](https://github.com/pinot-contrib/pinot-docs/pull/109)). Not sure why it hasn't been updated on documentation yet.

Oh i see. lemme check on that

**Comment by siddharthteotia** [comment]
@npawar - User docs were added some time ago by @SabrinaZhaozyf 

Just linking them here for reference - https://docs.pinot.apache.org/configuration-reference/functions/base64
